packages/nx/src/project-graph/build-dependencies/extract-file-references.ts

```diff
@@ -0,0 +1,488 @@
+        }
+
+        // Check if at least one of the alternatives has moved the scanner forwards
+        if (
+          tryConsumeImport() ||
+          tryConsumeExport() ||
+          tryConsumeRequireCall(
+            /*skipCurrentToken*/ false,
+            /*allowTemplateLiterals*/ true
+          )
+        ) {
+          continue;
+        } else {
+          nextToken();
+        }
+      }
+
+      scanner.setText(undefined);
+    }
+
+    processImports();
+
+    /**
+     * Extract any <reference path="..."> pragmas from the file
+     */
+    for (const range of ts.getLeadingCommentRanges(sourceText, 0) || []) {
+      const comment = sourceText.substring(range.pos, range.end);
+      const tripleSlash =
+        range.kind === SyntaxKind.SingleLineCommentTrivia &&
+        tripleSlashXMLCommentStartRegEx.exec(comment);
+      if (!tripleSlash) {
+        continue;
+      }
+      const pragmaName = tripleSlash[1].toLowerCase();
+      if (pragmaName !== 'reference') {
+        continue;
+      }
+      const matchResult = referencePathArgRegEx.exec(comment);
+      if (!matchResult) {
+        continue; // The <reference /> does not have a path argument, ignore it
+      }
+      const value = matchResult[2] || matchResult[3];
+      const startPos =
+        range.pos + matchResult.index + matchResult[1].length + 1;
+
+      allFileReferences.push({
+        pos: startPos,
+        end: startPos + value.length,
+        fileName: value,
+      });
+    }
+
+    /**
+     * We only need to build a full ts.SourceFile (AST) if at least one of the following is true:
+     *
+     * 1. The source contains one or more nx-ignore-next-line comments
+     * 2. The source contains legacy Angular loadChildren string syntax
+     */
+    let sourceFile: SourceFile | undefined;
+    let lineStarts: number[] | undefined;
+
+    const ignoredLines = new Set();
+
+    const totalNxIgnoreNextLineOccurrences =
+      sourceText.split(NX_IGNORE_NEXT_LINE_COMMENT).length - 1;
+    if (totalNxIgnoreNextLineOccurrences > 0) {
+      if (!sourceFile) {
+        sourceFile = getSourceFile(filePath, sourceText);
+      }
+      if (!lineStarts) {
+        lineStarts = computeLineStarts(sourceText);
+      }
+
+      forEachComment(sourceFile, (_, comment) => {
+        const commentTextContents = sourceText
+          .substring(
+            comment.pos + 2,
+            comment.kind === SyntaxKind.SingleLineCommentTrivia
+              ? comment.end
+              : comment.end - 2
+          )
+          .trim()
+          .toLowerCase();
+        if (commentTextContents !== NX_IGNORE_NEXT_LINE_COMMENT) {
+          return;
+        }
+
+        const endLineNumberOfNxIgnoreComment = computeLineOfPosition(
+          lineStarts,
+          comment.end
+        );
+
+        const ignoredLine = endLineNumberOfNxIgnoreComment + 1;
+        ignoredLines.add(ignoredLine);
+
+        allFileReferences = allFileReferences.filter((file) => {
```

Comment:
Good job trying optimize this...

Can't this be faster because we know they are both in order though?

After we find the references in a line, we can ignore references before those in subsequent lines. We should have a pointer to start checking file references after the ones we've already checked. And once we reach a file reference, that starts after the next line start, we should make that where we start? Does that make sense?

Reply:
This wasn't intended to be an optimization per se, this was just a functional implementation taking some inspiration from how it worked in `tslint` with their disable comments.

Your suggested optimization makes sense though!